crates/protocol/src/io.rs

```diff
@@ -0,0 +1,301 @@
+use std::borrow::Cow;
+use std::io::{Cursor, Read};
+use std::iter;
+use std::marker::PhantomData;
+
+use anyhow::{bail, Context};
+use byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
+use thiserror::Error;
+use uuid::Uuid;
+
+use crate::var_int::VarInt;
+
+pub trait Decoder: Sized {
+    fn decode(buffer: &mut Cursor<&[u8]>) -> anyhow::Result<Self>;
+}
+
+pub trait Encoder {
+    fn encode(&self, buffer: &mut Vec<u8>) -> anyhow::Result<()>;
+}
+
+macro_rules! integer_impl {
+    ($($int:ty, $read_fn:tt, $write_fn:tt),* $(,)?) => {
```

Comment:
Let's use a more line-independent grammar:

```suggestion
    ($([$int:ty, $read_fn:tt, $write_fn:tt]),* $(,)?) => {
```

which will be usable as:

```rust
integer_impl! {
    [u16, read_u16, write_u16],
    [u32, read_u32, write_u32],
    ...
}
```

also try to use more specific token types (`$tt` is the most generic one) for better compile-time errors, e.g.:

```suggestion
    ($([$int:ty, $read_fn:ident, $write_fn:ident]),* $(,)?) => {
```